File: 파이썬/판교가는길/dfs/안전 영역/Main2.py
import sys
sys.setrecursionlimit(10**6)
N = int(input())

_map = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]

maxHeight = max(max(_map))

dr = [1, -1, 0, 0]
dc = [0, 0, 1, -1]


def dfs(r, c, height):
    for i in range(4):
        nr = r + dr[i]
        nc = c + dc[i]
        if 0 <= nr < N and 0 <= nc < N and _map[nr][nc] > height and not checked[nr][nc]:
            checked[nr][nc] = True
            dfs(nr, nc, height)


ans = 0

for height in range(maxHeight):
    area = 0
    checked = [[False] * N for _ in range(N)]
    for r in range(N):
        for c in range(N):
            if not checked[r][c] and _map[r][c] > height:
                checked[r][c] = True
                dfs(r, c, height)
                area += 1
    ans = max(ans, area)

print(ans)

File: Programmers/src/바킹독/스택/히스토그램에서가장큰직사각형/Main2.java
package 바킹독.스택.히스토그램에서가장큰직사각형;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.Stack;

public class Main2 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] input;
        Stack<Integer> s = new Stack<>();
        StringBuilder sb = new StringBuilder();
        while (true) {
            input = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::valueOf).toArray();

            int n = input[0];

            if (n == 0) {
                break;
            }

            long[] arr = new long[n];

            for (int i = 0; i < n; i++) {
                arr[i] = input[i + 1];
            }

            s.clear();

            long ans = 0;

            for (int i = 0; i < n; i++) {
                long h = arr[i];
                while (!s.isEmpty() && arr[s.peek()] > h) {
                    int idx = s.pop();
                    long height = arr[idx];
                    int width = s.isEmpty() ? i : i - s.peek() - 1;

                    long area = height * width;

                    ans = Math.max(ans, area);
                }

                s.push(i);
            }

            while (!s.isEmpty()) {
                int idx = s.pop();
                long height = arr[idx];
                int width = s.isEmpty() ? n : n - s.peek() - 1;
                long area = height * width;

                ans = Math.max(ans, area);
            }

            sb.append(ans + "\n");
        }
        System.out.println(sb);
    }
}
